wfc.h

```diff
@@ -204,6 +205,28 @@ void wfc_destroy(struct wfc *wfc);
 
 #endif
 
+struct wfc__noise {
+  unsigned int seed;
+  unsigned int roll;
+};
+
+// SquirrelNoise3 function by Squirrel Eiserloh
+inline int wfc__noise_hash(struct wfc__noise* noise) {
```

Comment:
Oh I'm sorry, I didn't mean to submit that change, only the edits that solved my compiler errors. I added the custom rand for myself because I'm using your library to generate maps in an RTS game and I wanted to be sure of its determinism.

Reply:
I think this is a good change. I'd like to keep it. Determinism is useful so that you can generate the same results from the same seeds. And also it might be useful that we don't alter state of `rand()`, in case the user app uses it for something else.

The only thing that I'd suggest is that we pass `seed` as part of `wfc_run` instead of `wfc_init`, because `wfc_init` doesn't have to be called before the first run. Currently `wfc_run` takes two arguments: the `wfc` struct and `max_collapse_cnt`: https://github.com/krychu/wfc/blob/master/wfc.h#L1120

I think the latter is not really that useful, we should change it to `seed`. `-1` would use some random seed (`time(NULL)`), and any natural number would be used as the actual seed. How does this sound to you?

An alternative would be to introduce `wfc_seed(seed)` which can be called between the runs, but that seems more convoluted?

I'm not familiar with SquirrelNoise3, need to read about it. Looks very performant which is always very welcome, does it have some other good properties why you use it?

Thanks again!

Reply:
I'm also thinking this could be a #define switch to opt into custom rand implementation, wdyt?

Reply:
Hi, I'm glad you like my changes :> I've removed those constexprs now

The reason seed is passed to wfc_init (and wfc_overlapping) in my version is because in the original library, wfc_init is where srand is called. I don't want to pass the seed into wfc_run, I want to set the seed once and run the generation multiple times for different purposes. So if you really want to change it, I would say using `wfc_seed(seed)` is the most versatile, and it would also be backwards compatible with the existing API, which is nice.

> I'm also thinking this could be a #define switch to opt into custom rand implementation, wdyt?

Could do, I avoid putting code behind ifdefs where possible but it's your library and it doesn't bother me.

I didn't think too hard about the noise function. it's just the one I happened to know about, and the author says it's got a very even distribution, which is good enough for me.

By the way, is including float.h necessary for any reason?

Reply:
Thanks.

I missed that you also passed `seed` to `wfc_overlapping` in which case my argument doesn't make sense.

> I want to set the seed once and run the generation multiple times for different purposes

I see, but having the `seed` in `wfc_init` is not good then, because you have to call `wfc_init` before each run (if you're re-using the same `struct wfc`) and so you are forced to pass the `seed` for each run rather than once at the beginning. Is this right? or am I missing something?

What's nice about `wfc_srand` is what you said and also that you can essentially choose whether you want to call it before each run or once at the beginning. And also you can decide to not call it at all in which case we use `time(NULL)`.

The only thing I'm slightly concerned about is that the specific rand implementation could have big impact on the quality of results. I'll do some testing and comparisons. If we switch from std rand to squirrel or something else, we should be sure it's not leading to worse results.

Reply:
float.h is for DBL_MAX only, I created an issue to remove it: https://github.com/krychu/wfc/issues/26

Reply:
>I see, but having the seed in wfc_init is not good then, because you have to call wfc_init before each run (if you're re-using the same struct wfc) and so you are forced to pass the seed for each run rather than once at the beginning. Is this right? or am I missing something?

Nah, you're right, I want to change it now. I looked at my code, and I use wfc_overlapping each time. how cool.

> The only thing I'm slightly concerned about is that the specific rand implementation could have big impact on the quality of results. I'll do some testing and comparisons. If we switch from std rand to squirrel or something else, we should be sure it's not leading to worse results.

Okay, let me know!

Reply:
I created issue to work on the selection of the better noise(): https://github.com/krychu/wfc/issues/27

In this talk: https://www.youtube.com/watch?v=LWFzPP8ZbdU I found that the implementation is a bit different from yours:
<img width="1219" alt="Seeding Noise functions" src="https://user-images.githubusercontent.com/947457/201344672-2e22d1b9-6075-4948-a5e3-828d8d8191c7.png">

your impl:
<img width="285" alt="image" src="https://user-images.githubusercontent.com/947457/201344816-c15eacd1-e423-4f01-9cf7-e2c18392fec1.png">

Notably in the reference implementation the second noise is added rather than multiplied, and the second bit shift is to the left << rather than to the right >>.

Is this something you were experimenting with?

I also noticed there was some issue reported with this noise and author since improved the implementation. I'll paste more info in the issue.

Reply:
>Is this something you were experimenting with?

no, that's just a mistake.............................................................

Reply:
no worries, good to know. We're on a good track though. Next week I'll probably finish testing strongest candidates including squirrel 3 and 5. This has been very helpful thanks.